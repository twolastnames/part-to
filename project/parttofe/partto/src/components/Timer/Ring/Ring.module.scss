@import "../../../App.scss";

.label {
  width: 100%;
  display: flex;
  justify-content: center;
}

.ring {
  width: 100%;
  height: 100%;
}

.ringRoot {
  width: 100%;
  height: 100%;
}
