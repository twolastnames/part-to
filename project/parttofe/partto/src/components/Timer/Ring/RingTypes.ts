import { ReactNode } from "react";

export interface RingProps {
  magnitude: number;
  label: ReactNode;
}
