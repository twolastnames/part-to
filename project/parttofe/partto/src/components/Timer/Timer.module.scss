@import "../../App.scss";

@mixin label {
  font: var(--font-strong-xxxlarge);
}

.normal {
  @include label;
}

@keyframes blink {
  0% {
    opacity: 1;
  }
  50% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}

.overdue {
  @include label;
  animation: blink 1s infinite;
}
