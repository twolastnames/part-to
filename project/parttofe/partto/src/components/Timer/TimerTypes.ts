import { DateTime } from "../../shared/dateTime";
import { Duration } from "../../shared/duration";

export interface TimerProps {
  start?: DateTime;
  duration: Duration;
}
