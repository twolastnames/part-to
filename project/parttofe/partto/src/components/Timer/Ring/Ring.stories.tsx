import React from "react";
import type { Meta, StoryObj } from "@storybook/react";

import { Ring } from "./Ring";
import { ShellProvider } from "../../../providers/ShellProvider";

const meta: Meta<typeof Ring> = {
  component: Ring,
  decorators: (Story) => (
    <ShellProvider>
      <div style={{ width: "200px", height: "200px" }}>
        <Story />
      </div>
    </ShellProvider>
  ),
};
export default meta;

type Story = StoryObj<typeof Ring>;

export const Simple: Story = {
  args: {
    magnitude: 40,
    label: "world hello",
  },
};

export const First: Story = {
  args: {
    magnitude: 70,
    label: "hello world",
  },
};

export const Second: Story = {
  args: {
    magnitude: 160,
    label: "good bye",
  },
};

export const Third: Story = {
  args: {
    magnitude: 250,
    label: "bye good",
  },
};
