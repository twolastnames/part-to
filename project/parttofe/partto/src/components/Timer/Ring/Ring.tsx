import React, { useLayoutEffect, useRef, useState } from "react";

import classes from "./Ring.module.scss";
import { RingProps } from "./RingTypes";
import { RingProgress } from "@mantine/core";
import { useWindowResize } from "../../../hooks/windowResize";

export function Ring({ magnitude, label }: RingProps) {
  const { height, width } = useWindowResize();
  const [size /* TODO: reimplment here */] = useState<number>(1);
  const ringRef = useRef<HTMLDivElement | null>(null);

  const color1 =
    magnitude >= 1 ? "var(--highlight-color)" : "var(--secondary-color)";
  const value1 = magnitude >= 2 ? 1 : magnitude % 1;
  const color2 =
    magnitude >= 1 ? "var(--secondary-color)" : "var(--highlight-color)";
  const value2 = magnitude >= 2 ? 0 : 1 - (magnitude % 1);

  useLayoutEffect(() => {
    const currentSize = Math.min(
      ringRef.current?.offsetHeight || Infinity,
      ringRef.current?.offsetWidth || Infinity,
    );
    if (currentSize !== size) {
      //TODO: Fix rerendering issue here
      //setSize(currentSize);
    }
  }, [height, width, size]);

  return (
    <div ref={ringRef} className={classes.ring} data-testid="Ring">
      <RingProgress
        label={<div className={classes.label}>{label}</div>}
        size={size}
        thickness={8}
        sections={[
          { color: color1, value: value1 * 100 },
          { color: color2, value: value2 * 100 },
        ]}
      />
    </div>
  );
}
