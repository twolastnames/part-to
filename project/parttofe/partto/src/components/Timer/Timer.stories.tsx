import React from "react";
import type { Meta, StoryObj } from "@storybook/react";

import { Timer } from "./Timer";
import { ShellProvider } from "../../providers/ShellProvider";
import { getDateTime } from "../../shared/dateTime";
import { getDuration } from "../../shared/duration";

const meta: Meta<typeof Timer> = {
  component: Timer,
  decorators: (Story) => (
    <ShellProvider>
      <div style={{ width: "400px", height: "420px" }}>
        <Story />
      </div>
    </ShellProvider>
  ),
};
export default meta;

type Story = StoryObj<typeof Timer>;

export const Simple: Story = {
  args: {
    start: getDateTime(),
    duration: getDuration(65000),
  },
};

export const Hours: Story = {
  args: {
    start: getDateTime(),
    duration: getDuration(96000000),
  },
};
