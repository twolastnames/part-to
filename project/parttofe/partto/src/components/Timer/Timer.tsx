import React, { useEffect, useState } from "react";

import classes from "./Timer.module.scss";
import { TimerProps } from "./TimerTypes";
import { Ring } from "./Ring/Ring";
import { DateTime, getDateTime } from "../../shared/dateTime";
import { DurationFormat } from "../../shared/duration";

export function Timer({ start, duration }: TimerProps) {
  const [started] = useState<DateTime>(start || getDateTime());
  const [magnitude, setMagnitude] = useState<number>(0);

  useEffect(() => {
    const listener = () => {
      const passed = getDateTime().sinceEpoch() - started.sinceEpoch();
      setMagnitude(passed / duration.toMilliseconds());
    };
    const id = setInterval(listener, 1000);
    return () => {
      clearInterval(id);
    };
  }, [started, duration]);

  return (
    <Ring
      label={
        <div className={magnitude > 2 ? classes.overdue : classes.normal}>
          {started
            .add(duration)
            .subtract(getDateTime())
            .format(DurationFormat.TIMER)}
        </div>
      }
      magnitude={magnitude}
    />
  );
}
