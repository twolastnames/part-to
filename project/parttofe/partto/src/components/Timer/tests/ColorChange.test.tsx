import React from "react";
import { expect, it, describe } from "@jest/globals";
import { render, screen } from "@testing-library/react";
import { ShellProvider } from "../../../providers/ShellProvider";
import { Timer } from "../Timer";
import { getDuration } from "../../../shared/duration";
import { RingedRing } from "../Ring/Ring";

describe("color change", () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });
  afterAll(() => {
    jest.useRealTimers();
  });
  it("will change", () => {
    render(
      <ShellProvider>
        <Timer duration={getDuration(5000)} ringClasses={RingedRing} />
      </ShellProvider>,
    );
    jest.advanceTimersByTime(7000);
    const component = screen.getByTestId("Ring");
    expect(component).toMatchSnapshot();
  });
});
