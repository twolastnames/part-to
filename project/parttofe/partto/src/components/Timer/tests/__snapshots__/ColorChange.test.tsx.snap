// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`color change will change 1`] = `
<div
  class="ring"
  data-testid="Ring"
>
  <div
    class="m_b32e4812 mantine-RingProgress-root"
    style="--rp-size: calc(Infinityrem * var(--mantine-scale)); --rp-label-offset: calc(1rem * var(--mantine-scale));"
  >
    <svg
      class="m_d43b5134 mantine-RingProgress-svg"
    >
      <circle
        class="m_b1ca1fbf mantine-RingProgress-curve"
        cx="Infinity"
        cy="Infinity"
        fill="none"
        r="Infinity"
        stroke-dasharray="NaN, Infinity"
        stroke-dashoffset="0"
        stroke-linecap="butt"
        stroke-width="8"
        transform="scale(1, -1) translate(0, -Infinity)"
      />
      <circle
        class="m_b1ca1fbf mantine-RingProgress-curve"
        cx="Infinity"
        cy="Infinity"
        fill="none"
        r="Infinity"
        stroke-dasharray="NaN, Infinity"
        stroke-dashoffset="Infinity"
        stroke-linecap="butt"
        stroke-width="8"
        style="--curve-color: var(--secondary-color);"
      />
      <circle
        class="m_b1ca1fbf mantine-RingProgress-curve"
        cx="Infinity"
        cy="Infinity"
        fill="none"
        r="Infinity"
        stroke-dasharray="Infinity, NaN"
        stroke-dashoffset="0"
        stroke-linecap="butt"
        stroke-width="8"
        style="--curve-color: var(--highlight-color);"
      />
    </svg>
    <div
      class="m_b23f9dc4 mantine-RingProgress-label"
    >
      <div
        class="label"
      >
        <div
          class="normal"
        >
          5
        </div>
      </div>
    </div>
  </div>
</div>
`;
